File: python/hashing/internationalPhoneBookLinkedList.py
class InternationalPhoneBookLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None

    class Node:
        def __init__(self, number, name):
            self.number = number
            self.name = name
            self.next = None
            self.prev = None

    def setName(self, number, name):
        node = self.Node(number, name)

        if self.head is None:
            self.head = node
        
        if self.tail is not None:
            self.tail.next = node
            node.prev = self.tail

        self.tail = node

    def getName(self, number):
        node = self.head

        while node is not None:
            if node.number == number:
                return node.name
            node = node.next

        return None

File: java/src/hashing/InternationalPhoneBookLinkedList.java
package hashing;

public class InternationalPhoneBookLinkedList {
    private Node head;

    private Node tail;

    private class Node {
        private String number;

        private String name;

        private Node next;

        private Node prev;

        public Node(String number, String name) {
            this.number = number;
            this.name = name;
        }
    }

    public void setName(String number, String name) {
        Node node = new Node(number, name);

        if (this.head == null)
            this.head =  node;

        if (this.tail != null) {
            this.tail.next = node;
            node.prev = this.tail;
        }

        this.tail = node;
    }

    public String getName(String number) {
        Node node = this.head;

        while (node != null) {
            if (node.number == number)
                return node.name;
            node = node.next;
        }

        return null;
    }
}
